### PR TITLE
Feature user crud

### PR DESCRIPTION
**Pull request by Bruh3509**
Дописал CRUD, добавил testcontainers и написал немного тестов к service (потом еще допишу).

### CODE REVIEW

File: src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        return mapper.entityToDto(userRepository.save(user));
+    }
+
+    @Override
+    public void deleteUser(Long id) throws UserNotFoundException {
+        var user = userRepository.findById(id).orElseThrow(UserNotFoundException::new); // а нужно ли исключение?
+        userRepository.deleteById(id);
+    }
+
+    private void checkUniqueness(UserDto userDto) throws UniqueRecordException {
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java]: Here i suggest you to use Strategy pattern or switch. And create utility class for static String

File: src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        // validation
+        checkUniqueness(userDto);
+
+        var user = mapper.dtoToEntity(userDto);
+        return mapper.entityToDto(userRepository.save(user));
+    }
+
+    @Override
+    public void deleteUser(Long id) throws UserNotFoundException {
+        var user = userRepository.findById(id).orElseThrow(UserNotFoundException::new); // а нужно ли исключение?
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java]: You can create two methods: Optional<User>findSafe(Long id) and User getById(Long id){
//Pls after second dot push enter and use convention like here:
userRepository.findById(id)
.orElseThrow(UserNotFoundException::new);
}

File: src/test/java/com/demo/cars/database/DbTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.annotation.Rollback;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SpringBootTest
+@Transactional
```
Comment (Mishgan55) [inline src/test/java/com/demo/cars/database/DbTest.java]: Тесты переписать использовать Mockito

File: src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return mapper.entityToDtoList(userRepository.findAll());
+    }
+
+    @Override
+    public UserDto getUserById(Long id) {
+        var optUser = userRepository.findById(id);
+        if (optUser.isPresent()) {
+            return mapper.entityToDto(optUser.get());
+        } else {
+            return new UserDto(); // если не найден user,
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java]: Definitely throw exception. You can find solution below

File: src/main/java/com/demo/cars/database/service/UserService.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+import com.demo.cars.database.exception.UniqueRecordException;
+import com.demo.cars.database.exception.UserNotFoundException;
+import com.demo.cars.dto.UserDto;
+
+import java.util.List;
+
+// User(entity) or UserDto
+public interface UserService {
+    // договорились возвращать dto, но есть ли смысл тут, если и так передаем что возвращаем
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/service/UserService.java]: UserDto saveUser(Long id, UserRequset userRequest)

File: src/main/java/com/demo/cars/database/service/UserService.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    void regUser(UserDto userDto) throws UniqueRecordException;
+    List<UserDto> getAllUsers();
+    UserDto getUserById(Long id);
+    UserDto updateUser(Long id, UserDto userDto) throws UserNotFoundException, UniqueRecordException;
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/service/UserService.java]: UserDto updateUser(Long id, UserRequest userRequest) and don't use throw here

File: src/main/java/com/demo/cars/database/fields/user/UserColumn.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.demo.cars.database.fields.user;
+
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/fields/user/UserColumn.java]: For what?

File: src/main/java/com/demo/cars/database/exception/UniqueRecordException.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package com.demo.cars.database.exception;
+
+import lombok.AccessLevel;
+import lombok.experimental.FieldDefaults;
+
+@FieldDefaults(level = AccessLevel.PUBLIC, makeFinal = true) // хотел еще добавить сюда же в аннотации accessLevel = Public, но что-то ругался, якобы не Public и не видно
+public class UniqueRecordException extends Exception {
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/database/exception/UniqueRecordException.java]: ou should create Utill package and there create PropertyUtill for static fields

**Review by Mishgan55 — CHANGES_REQUESTED** [review]

File: src/test/java/com/demo/cars/database/DbTest.java (hunk 3 above)
Comment (Bruh3509) [reply src/test/java/com/demo/cars/database/DbTest.java]: Окей, мне правда не совсем понятно, что тут мокать конкретно и для чего, если у меня база и все остальное крутиться в testcontainers?

File: src/main/java/com/demo/cars/database/service/UserService.java (hunk 5 above)
Comment (Bruh3509) [reply src/main/java/com/demo/cars/database/service/UserService.java]: UserRequest это от контроллера приходить должно?

File: src/main/java/com/demo/cars/database/service/UserService.java (hunk 6 above)
Comment (Bruh3509) [reply src/main/java/com/demo/cars/database/service/UserService.java]: How can I manage unique record and user not found issues then? (if I am not throwing exceptions?)

File: src/main/java/com/demo/cars/database/fields/user/UserColumn.java (hunk 7 above)
Comment (Bruh3509) [reply src/main/java/com/demo/cars/database/fields/user/UserColumn.java]: forgot to delete

File: src/main/java/com/demo/cars/database/exception/UniqueRecordException.java (hunk 8 above)
Comment (Bruh3509) [reply src/main/java/com/demo/cars/database/exception/UniqueRecordException.java]: Ok. I just thought that message Strings to Exception should/can be in the Exception Class

File: src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java (hunk 1 above)
Comment (Bruh3509) [reply src/main/java/com/demo/cars/database/service/impl/UserServiceImpl.java]: I was thinking about switch but didn't find out how can I use it here. I will try the Strategy pattern then

File: src/main/java/com/demo/cars/CarsApplication.java [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/CarsApplication.java]: I think you forgot to remove some imports, i suggest you to install plugin Sonarlint

File: src/main/java/com/demo/cars/service/UserService.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package com.demo.cars.service;
+
+import com.demo.cars.dto.UserDto;
+
+import java.util.List;
+
+public interface UserService {
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/service/UserService.java:7]: remove this comments

File: src/main/java/com/demo/cars/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import com.demo.cars.mapper.UserMapper;
+import jakarta.transaction.Transactional;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.FieldDefaults;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+import static com.demo.cars.utility.PropertyUtil.*;
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/service/impl/UserServiceImpl.java:17]: Don't use wildcards for imports

File: src/main/java/com/demo/cars/entity/User.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
+package com.demo.cars.entity;
+
+import jakarta.persistence.*;
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/entity/User.java:3]: Don't use wildcards for imports

File: src/main/java/com/demo/cars/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+@Transactional
+@RequiredArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+public class UserServiceImpl implements UserService {
+    UserRepository userRepository;
+    UserMapper mapper;
+
+    @Override
+    public void regUser(UserDto userDto) {
+        // validation
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/service/impl/UserServiceImpl.java]: delete comment

File: src/main/java/com/demo/cars/service/impl/UserServiceImpl.java (hunk 11 above)
Comment (Bruh3509) [reply src/main/java/com/demo/cars/service/impl/UserServiceImpl.java:17]: It's auto refactoring from ide I guess

File: src/test/java/com/demo/cars/database/UserServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+
+    @BeforeEach
+    void resetMockRules() {
+        Mockito.reset(userRepository);
+        Mockito.reset(userMapper);
+    }
+
+    @Test
+    void testFindAll() {
+        // arrange
```
Comment (Mishgan55) [inline src/test/java/com/demo/cars/database/UserServiceTest.java:44]: Please use when method, here is example:
        User user = createUser();
        UserDto userDto = createUserDto();

        // act
        when(userRepository.findAll()).thenReturn(List.of(user));
        when(userMapper.entityToDtoList(List.of(user))).thenReturn(List.of(userDto));
        var result = userService.getAllUsers();

        // assert
        assertEquals(1, result.size());
        assertEquals("JASO1921AS", result.getFirst().getPassportId());

File: src/main/java/com/demo/cars/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+                .map(mapper::entityToDto)
+                .orElseThrow(UserNotFoundException::new);
+    }
+
+    @Override
+    public UserDto updateUser(Long id, UserDto userDto) {
+        userRepository.findById(id)
+                .orElseThrow(UserNotFoundException::new);
+
+        // validation
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/service/impl/UserServiceImpl.java]: delete comment

File: src/main/java/com/demo/cars/utility/PropertyUtil.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package com.demo.cars.utility;
+
+public final class PropertyUtil {
+    public static String EMAIL_EXC = "email unique exception";
```
Comment (Mishgan55) [inline src/main/java/com/demo/cars/utility/PropertyUtil.java]: better write full name UNIQUE_EXCEPTION

**Pull request merged**